### PR TITLE
Infrastructure for running Get Tests in C# V1 and Beta

### LINKED ISSUE #143: Create projects for executing snippets
Create V1 and Beta projects for executing Get tests. 
[AB#7619](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/7619)

### LINKED ISSUE #146: Create pipelines for running get tests
Scope: C# V1 and Beta 
[AB#7622](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/7622)

### PR DESCRIPTION
**Pull request by zengin**
- renamed test run name for compilation tests from `Run` to `Compile`
- introduced a new test type called `Execute`
- created two new projects that call `Execute` (C# V1 and Beta)
- `Execute` test does the following:
  - changes snippet decoration to extract request url for logging purposes
  - creates an Assembly from `Compile` phase
  - creates a ClientCredentialProvider using secrets from `appsettings.json`
  - calls the main method in the generated Assembly using the auth provider
  - if no exception is thrown, the test passes
  - if an exception is thrown, it is logged with the compiled snippet
- added two new pipelines for executing these tests:
  -  [V1 C# Snippet Execution Tests](https://microsoftgraph.visualstudio.com/Graph%20Developer%20Experiences/_build/results?buildId=38054&view=ms.vss-test-web.build-test-results-tab)
  -  [Beta C# Snippet Execution Tests](https://microsoftgraph.visualstudio.com/Graph%20Developer%20Experiences/_build/results?buildId=38050&view=ms.vss-test-web.build-test-results-tab)
  - Pipelines use secret variables in Azure DevOps and they are fed in to `appsettings.json` file using `scripts/transformAppSettings.ps1` script.
- Test data selection logic ignores known issues of compilation as it doesn't make sense to run those tests for execution (i.e. compilation is a prerequisite for execution)
- added @finsharp as code owner

Note: This PR doesn't address the delegated permissions scenario, it only uses application permissions. Delegated permissions will be addressed with a separate PR.
Fixes #143
Fixes #146

### CODE REVIEW

File: azure-pipelines/common-templates/transform-app-settings.yml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+steps:
+  - pwsh: '$(Build.SourcesDirectory)/msgraph-sdk-raptor/scripts/transformAppSettings.ps1 -TenantID $(TenantID) -ClientID $(ClientID) -ClientSecret $(ClientSecret)'
```
Comment (baywet) [inline azure-pipelines/common-templates/transform-app-settings.yml:2]: nit: I know the work is already done but: couldn't we have used environment variables instead?

File: scripts/transformAppSettings.ps1 [hunk 2]
```diff
@@ -0,0 +1,23 @@
+param(
+    [bool]$IsLocalRun = $false,
+    [bool]$GenerateLinqPadOutputInLocalRun = $false,
+    [string]$DocsRepoCheckoutDirectory = "C:/github",
+    [string]$TenantID,
```
Comment (baywet) [inline scripts/transformAppSettings.ps1:5]:
```suggestion
   [Parameter(Mandatory=$true)][string]$TenantID,
```

File: scripts/transformAppSettings.ps1 [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    [string]$ClientID,
```
Comment (baywet) [inline scripts/transformAppSettings.ps1:6]:
```suggestion
    [Parameter(Mandatory=$true)][string]$ClientID,
```

File: scripts/transformAppSettings.ps1 [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    [string]$ClientSecret
```
Comment (baywet) [inline scripts/transformAppSettings.ps1:7]:
```suggestion
    [Parameter(Mandatory=$true)][string]$ClientSecret
```

File: TestsCommon/CSharpTestRunner.cs [hunk 5]
```diff
@@ -75,6 +72,105 @@ public static void Run(LanguageTestData testData)
...
+            }
+        }
+
+        private static string CaptureUriAndHeadersInException(string codeToCompile)
+        {
+            var resultVariablePattern = "var ([a-zA-Z0-9]+) = await graphClient";
+            string resultVariable = null;
+            try
+            {
+                resultVariable = Regex.Match(codeToCompile, resultVariablePattern).Groups[1].Value;
```
Comment (baywet) [inline TestsCommon/CSharpTestRunner.cs:147]: static field for the regex instead? this way it gets compiled only once?

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs [hunk 6]
```diff
@@ -68,28 +74,99 @@ public CompilationResultsModel CompileSnippet(string codeSnippet, Versions versi
...
+                    var clientSecret = config.GetSection("ClientSecret").Value;
+                    IConfidentialClientApplication confidentialClientApp = ConfidentialClientApplicationBuilder
+                        .Create(clientId)
+                        .WithTenantId(tenantId)
+                        .WithClientSecret(clientSecret)
+                        .Build();
+                    authProvider = new ClientCredentialProvider(confidentialClientApp, "https://graph.microsoft.com/.default");
+
+                    var task = instance.Main(authProvider) as Task;
+                    task.Wait();
```
Comment (baywet) [inline msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:135]: any specific reason why this is not using the async/await syntax?

**Review by baywet — CHANGES_REQUESTED** [review]

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs [hunk 7]
```diff
@@ -68,28 +74,99 @@ public CompilationResultsModel CompileSnippet(string codeSnippet, Versions versi
...
+        /// <summary>
+        ///     Gets the result of the Compilation.Emit method.
+        /// </summary>
+        /// <param name="compilation">Immutable respresentation of a single invocation of the compiler</param>
+        private (EmitResult, Assembly) GetEmitResult(CSharpCompilation compilation)
+        {
+            Assembly assembly = null;
+
+            using MemoryStream memoryStream = new MemoryStream();
+            EmitResult emitResult = compilation.Emit(memoryStream);
```
[the remark this answers is not in the document]
Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:160]: Debugging the snippets is out of scope for this PR.

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs [hunk 8]
```diff
@@ -68,28 +74,99 @@ public CompilationResultsModel CompileSnippet(string codeSnippet, Versions versi
...
             {
                 metadataReferences.Add(MetadataReference.CreateFromFile(Path.Combine(graphAssemblyPathBeta, "Microsoft.Graph.Beta.dll")));
             }
-           
+
             var compilation = CSharpCompilation.Create(
                assemblyName,
                syntaxTrees: new[] { syntaxTree },
                references: metadataReferences,
                options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary));
```
[the remark this answers is not in the document]
Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:90]: Thanks for exploring the options here. Default is debug, I will set Release explicitly as we don't have the intentions to debug (at least not yet ;))

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs (hunk 6 above)
Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:135]: yes, I would like to execute the snippet in this try catch block. Adding async/await bubbles up the exception to the caller and handling the exception there requires a lot of refactoring.

We don't have a huge perf hit as all the tests are running parallel and this is very close to the end of a test execution (i.e. test result needs to wait on it anyways).

File: azure-pipelines/common-templates/transform-app-settings.yml (hunk 1 above)
Comment (zengin) [reply azure-pipelines/common-templates/transform-app-settings.yml:2]: I have multiple tenants and I use this script locally as well. It is easier to pass in parameters than setting env variables there.

File: TestsCommon/TestDataGenerator.cs [hunk 9]
```diff
@@ -468,6 +468,32 @@ public static class TestDataGenerator
...
+        /// </returns>
+        public static IEnumerable<TestCaseData> GetExecutionTestData(RunSettings runSettings)
+        {
+            return from testData in GetLanguageTestData(runSettings)
+                   let fullPath = Path.Join(GraphDocsDirectory.GetSnippetsDirectory(testData.Version, runSettings.Language), testData.FileName)
+                   let fileContent = File.ReadAllText(fullPath)
+                   let executionTestData = new ExecutionTestData(testData, fileContent)
+                   where !testData.IsKnownIssue // select compiling tests
+                   && fileContent.Contains("GetAsync()") // select only the get tests
+                   select new TestCaseData(executionTestData).SetName(testData.TestName).SetProperty("Owner", testData.TestName);
```
[the remark this answers is not in the document]
Comment (zengin) [reply TestsCommon/TestDataGenerator.cs:493]: @finsharp, I looked at this, but the source of categorization is not something consistent in the documentation link. If you find a better source for it, please let me know. Some counter examples:
```
timeoff-put
schedulinggroup-put
site-follow
```

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs [hunk 10]
```diff
@@ -68,28 +74,100 @@ public CompilationResultsModel CompileSnippet(string codeSnippet, Versions versi
...
+                    IAuthenticationProvider authProvider;
+
+                    var tenantId = config.GetSection("TenantID").Value;
+                    var clientSecret = config.GetSection("ClientSecret").Value;
+                    IConfidentialClientApplication confidentialClientApp = ConfidentialClientApplicationBuilder
+                        .Create(clientId)
+                        .WithTenantId(tenantId)
+                        .WithClientSecret(clientSecret)
+                        .Build();
+                    authProvider = new ClientCredentialProvider(confidentialClientApp, "https://graph.microsoft.com/.default");
```
Comment (nikithauc) [inline msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:133]: Can a constant be used for the scope?

**Review by nikithauc — APPROVED** [review]

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs (hunk 6 above)
Comment (baywet) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:136]: I'm not sure I'm following, there's already a try catch surrounding the block. If you change to async await, and change the type of the exception being caught in the catch right below, won't that do the trick?

File: azure-pipelines/common-templates/transform-app-settings.yml (hunk 1 above)
Comment (baywet) [reply azure-pipelines/common-templates/transform-app-settings.yml:2]: locally: $env:MY_TENANT_ID = "id", CI use the env node. It ain't that hard :D but sure, I get it's a preference thing as well.

File: scripts/transformAppSettings.ps1 [hunk 11]
```diff
@@ -0,0 +1,23 @@
+param(
+    [bool]$IsLocalRun = $false,
+    [bool]$GenerateLinqPadOutputInLocalRun = $false,
+    [string]$DocsRepoCheckoutDirectory = "C:/github",
+   [Parameter(Mandatory=$true)][string]$TenantID,
```
Comment (baywet) [inline scripts/transformAppSettings.ps1:5]: woot! I caught @zengin making a space mistake !!!! 🎉🎉🎉🎉🎉🎉

Comment (baywet) [reply scripts/transformAppSettings.ps1:5]: my bad, it's actually my suggestion that introduced it. I'll celebrate another time then.

Comment (baywet) [reply scripts/transformAppSettings.ps1:5]:
```suggestion
    [Parameter(Mandatory=$true)][string]$TenantID,
```

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs (hunk 6 above)
Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs:143]: I got confused. I was trying to figure out why it was bubbling up and it was in front of me the whole time: the specific type in the catch block. Thanks for highlighting that! Will push the fix in a few moments.

Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs]: fixed

File: msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs (hunk 10 above)
Comment (zengin) [reply msgraph-sdk-raptor-compiler-lib/MicrosoftGraphCSharpCompiler.cs]: fixed

**Review by finsharp — APPROVED** [review]

**Review by baywet — APPROVED** [review]

**Pull request merged**